File: typescipt/src/compiler/cSharp/expressionStatementExceptions/simpleLexyFunctionFunctionExpressionStatementException.ts


export class SimpleLexyFunctionFunctionExpressionStatementException extends IExpressionStatementException {
   public matches(expression: Expression): boolean {
     return expression is FunctionCallExpression functionCallExpression
        && functionCallExpression.ExpressionFunction is LexyFunction;
   }

   public callExpressionSyntax(expression: Expression, context: ICompileFunctionContext): Array<StatementSyntax> {
     if (!(expression is FunctionCallExpression functionCallExpression))
       throw new Error(`expression should be FunctionCallExpression`);
     if (!(functionCallExpression.ExpressionFunction is LexyFunction lexyFunction))
       throw new Error(
         `functionCallExpression.ExpressionFunction should be ExtractResultsFunction`);

     let parameterVariable = $`{LexyCodeConstants.ParameterVariable}{Guid.NewGuid():N}`;
     let resultsVariable = $`{LexyCodeConstants.ResultsVariable}{Guid.NewGuid():N}`;

     let result = new Array<StatementSyntax>();
     result.AddRange(FillFunctionExpressionStatementException.FillStatementSyntax(
       parameterVariable,
       lexyFunction.FunctionParametersType,
       lexyFunction.MappingParameters));

     result.Add(RunFunction(lexyFunction, parameterVariable, resultsVariable));

     result.AddRange(ExtractFunctionExpressionStatementException.ExtractStatementSyntax(
       lexyFunction.MappingResults,
       resultsVariable));

     return result;
   }

   private static StatementSyntax RunFunction(LexyFunction lexyFunction, string parameterVariable,
     string resultsVariable) {
     let initialize = LexyFunctionCall.RunFunction(lexyFunction.FunctionName, parameterVariable);

     let variable = VariableDeclarator(Identifier(resultsVariable))
       .WithInitializer(EqualsValueClause(initialize));

     return LocalDeclarationStatement(
       VariableDeclaration(Types.Syntax(lexyFunction.FunctionResultsType))
         .WithVariables(SingletonSeparatedList(variable)));
   }
}

File: src/Lexy.Compiler/Compiler/CSharp/ExpressionStatementExceptions/SimpleLexyFunctionFunctionExpressionStatementException.cs
using System;
using System.Collections.Generic;
using Lexy.Compiler.Compiler.CSharp.BuiltInFunctions;
using Lexy.Compiler.Language.Expressions;
using Lexy.Compiler.Language.Expressions.Functions;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;

namespace Lexy.Compiler.Compiler.CSharp.ExpressionStatementExceptions;

internal class SimpleLexyFunctionFunctionExpressionStatementException : IExpressionStatementException
{
    public bool Matches(Expression expression)
    {
        return expression is LexyFunction;
    }

    public IEnumerable<StatementSyntax> CallExpressionSyntax(Expression expression)
    {
        if (expression is not FunctionCallExpression functionCallExpression)
        {
            throw new InvalidOperationException("expression should be FunctionCallExpression");
        }
        if (functionCallExpression is not LexyFunction lexyFunction) {
            throw new InvalidOperationException(
                "functionCallExpression.FunctionCallExpression should be ExtractResultsFunction");
        }

        var parameterVariable = $"{LexyCodeConstants.ParameterVariable}_{expression.Reference.LineNumber}";
        var resultsVariable = $"{LexyCodeConstants.ResultsVariable}_{expression.Reference.LineNumber}";

        var result = new List<StatementSyntax>();
        result.AddRange(FillFunctionExpressionStatementException.FillStatementSyntax(
            parameterVariable,
            lexyFunction.FunctionParametersType,
            lexyFunction.MappingParameters));

        result.Add(RunFunction(lexyFunction, parameterVariable, resultsVariable));

        result.AddRange(ExtractFunctionExpressionStatementException.ExtractStatementSyntax(
            lexyFunction.MappingResults,
            resultsVariable));

        return result;
    }

    private static StatementSyntax RunFunction(LexyFunction lexyFunction, string parameterVariable,
        string resultsVariable)
    {
        var initialize = LexyFunctionCall.RunFunction(lexyFunction.FunctionName, parameterVariable);

        var variable = VariableDeclarator(Identifier(resultsVariable))
            .WithInitializer(EqualsValueClause(initialize));

        return LocalDeclarationStatement(
            VariableDeclaration(Types.Syntax(lexyFunction.FunctionResultsType))
                .WithVariables(SingletonSeparatedList(variable)));
    }
}